cli `version`: Show date as part of the version next to the commit id, shorten commit id

ilyagr opened the pull request:
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the branch rather than adding commits on top. Use force-push when
pushing the updated branch (`jj git push` does that automatically when you
rewrite a branch). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [x] I have added tests to cover my changes

ilyagr commented:
We discussed the Nix part with @Ralith. He suggested

```nix
preConfigure = ''
    NIX_JJ_GIT_HASH=$(git <something> --git-dir=${./.}/jj/repo/store/git)
'';
```

I might look into that, or I could leave it for some more Nix person to do.

thoughtpolice commented:
Something like this should work, I think, in `flake.nix`, but we should change the date format from `git log` to be `YYYYMMDD` to be consistent::

```
gitRev = "${builtins.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}|${self.shortRev or "dirty"}";
```

thoughtpolice commented:
Commit cf39adafac5ed1f5adc112ede325d6fcbfd7d2b7 on branch `push-numyszkprmty` contains the changes I think you need. Here's an example of what this looks like when using the interactive shell, basically the format `<YYYYMMDD>|<HASH>{-dirty}`

```
austin@GANON:~/src/jj$ echo $NIX_JJ_GIT_HASH 
20230811|cf39adafac5ed1f5adc112ede325d6fcbfd7d2b7
```

```
austin@GANON:~/src/jj$ echo $NIX_JJ_GIT_HASH 
20230811|17b45d642f39351500d74bc07fe1d70ee0f7bb5c-dirty
```

The first is a clean commit and the second is dirty. This uses the `dirtyRev` and `dirtyShortRev` features from Nix that are only available in Nix as of a month or so ago, so it's *very* recent; on older versions of Nix, for backwards compatibility the version in a dirty repository will look like this:

```
austin@GANON:~/src/jj$ echo $NIX_JJ_GIT_HASH 
20230811|dirty
```

Which isn't useful, but mostly only happens to us during development, so I think it's OK.

If you can unify these two version paths to be the same, that would be great. Feel free to mold this commit to your needs.

Ralith commented:
> self.lastModifiedDate or self.lastModified

Won't this be different than the commit date?

thoughtpolice commented:
`lastModifiedDate` is [the commit time of the `rev` in question](https://github.com/NixOS/nix/blob/a1fdc68c655e8d7ece51b6c84796d35203386c87/src/nix/flake.md?plain=1#L379). (I think it would have to be, otherwise it would break determinism.)

necauqua commented:
The point of `NIX_JJ_GIT_HASH` was that nix cleans up the .git folder and `build.rs` is not run in the jj/git repo, so it couldn't load the hash.

I now realised that a better approach would probably be to set `NIX_GIT_REPO="${./.}/jj/repo/store/git"` or something similar, and have `--git-dir` set to that in `build.rs` - that way there is no two separate ways (in nix and in rust) to calculate this same bleeding version value

ilyagr commented:
@thoughtpolice Thanks! I don't feel like adding that to this PR *right now*, but it looks like it'd be easy to add. Either the PR can wait or it can go into a separate one.

I think adding "-dirty" should be a separate change, as it'd require adjusting a test, and ideally it'd go together with supporting it in the regular Rust code.

> I now realised that a better approach would probably be to set NIX_GIT_REPO="${./.}/jj/repo/store/git" or something similar, 

This sounds fine to me, but I don't grok Nix philosophy or implementation enough to really know whether this is a good idea. I think it should be a separate discussion/PR in any case.

ilyagr commented:
Also, I'll leave it up to you to decide which date is exactly right. It would be nice to be consistent, but we can always fix it later.

ilyagr commented:
I changed the date format to YYYYMMDD and incorporated some of @thoughtpolice's changes. Thank you!

I rewrote them a bit and removed the portion having to do with dirty revisions for now. We should add it back when the rest of `build.rs` understands dirty revisions.

ilyagr commented on flake.nix:
```diff
@@ -18,6 +18,11 @@
+      # TODO(aseipp): Use dirtyRev and dirtyShortRev to record dirty checkout
+      # when we update `build.rs` to understand dirty checkouts.
+      gitRev = self.rev or "";
+      gitDate = builtins.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101");
```
I'd prefer the date to be "" on failure, but I'm not sure whether `builtins.substring` will panic in that case. Since `gitRev` will be "", and `build.rs` recognizes that as a failure, it doesn't really matter.

thoughtpolice replied:
`builtins.substring` is basically just a way of taking a slice of another string; in the case of `""` it would just return `""` no matter the (positive) input ranges you provide.

ilyagr replied:
Thank you!

ilyagr commented:
There's a mild inconsistency in dates. I'm guessing one is in UTC and one is in the local timezone. I think that's fine, though if somebody wants to ensure UTC everywhere it'd probably be nicer. The UTC timestamp feature mentioned suggested in https://discord.com/channels/968932220549103686/969829516539228222/1139656152926404649 would be helpful here.

![image](https://github.com/martinvonz/jj/assets/4123047/4b7d77df-8c8c-4d1d-a8a6-b07d26eb64b2)

thoughtpolice commented:
FWIW, if you can set `TZ=UTC0` in the environment on the invocation to `git`, it should handle that fine on both Windows and Unix. This works on my Windows machine, for example:

```
PS C:\code\thingy> git show --quiet --date=local --format="%cd"
Tue Apr 16 02:17:33 2019
PS C:\code\thingy> $env:TZ = "UTC0"; git show --quiet --date=local --format="%cd"
Tue Apr 16 07:17:33 2019
PS C:\code\thingy> $env:TZ = "UTC0"; git show --quiet --date='format-local:%Y%m%dT%H%M%SZ' --format="%cd"
20190416T071733Z
```

ilyagr commented:
> FWIW, if you can set `TZ=UTC0` in the environment on the invocation to `jj`, it _might_ handle that fine on both Windows and Unix? It's POSIX specified, at least.

Thanks, that sounds good. I don't think I'll use it in this commit, though. It's blocked on the jj side by #2080, I think.

ilyagr commented:
I'll switch all the times to UTC in the near future. I have a draft at https://github.com/ilyagr/jj/commit/630048ac6a475936120cb1cd177d5daf564f26c6, but I think I'll change it up a little (see the second point below).

@thoughtpolice, two thoughts:

1. If I understand correctly from https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake.html#flake-format (and if that's the correct reference), `self.lastModified` is a number of seconds since the epoch, not a date.
2. Thanks for checking how to get git to print UTC dates. It seems a bit tenuous though -- even though you checked that it seems to work on Windows, the docs don't seem to suggest that it *has* to work on Windows. I'm worried that this might depend on how you compile git.

    So, now I'm thinking of switching to having Git print the timestamp in seconds since the (UTC) epoch (just like `self.lastModified`) and then converting it to a date in the Rust code.

ilyagr commented:
I updated this to use UTC date everywhere. Also, after I read www.semver.org, I changed the build data to be shown after a `+` in the version instead of a `-`. So, `jj version` now outputs in the following format:

    jj 0.10.0+20231027-bb8af5adc4f98ff2

thoughtpolice commented on flake.nix:
```diff
@@ -132,6 +138,8 @@
           export ZSTD_SYS_USE_PKG_CONFIG=1
           export LIBSSH2_SYS_USE_PKG_CONFIG=1
+          export NIX_JJ_GIT_HASH="${gitRev}"
+          export NIX_JJ_GIT_TIMESTAMP="${gitTimestamp}";
```
So this will ensure `NIX_JJ_GIT_TIMESTAMP` (and `NIX_JJ_GIT_HASH`) only actually gets configured when the environment is populated, which typically happens only when you run `nix develop` for the first time. I think this is probably OK, to be honest, since it's only a minor glitch when nix-oriented developers are hacking. It just means the `--version` will be wrong, but that's probably OK in the working copy.

To fix this we'd need to have a third case where we set something like `NIX_JJ_SHELL=1` and then have `build.rs` handle that case like the normal one where we just interrogate the `.git` dir.

ilyagr replied:
Ideally, these variables would only populate in the case when jj is compelled outside a git repo by nix.

It's beyond the scope of my comfort level with nix, though. Somebody else would have to test these things and make another PR, I think.

jyn514 commented on cli/build.rs:
```diff
@@ -41,21 +44,51 @@ fn main() -> std::io::Result<()> {
+
+/// Return the UTC committer date (maybe) and the git hash
+fn get_git_timestamp_and_hash() -> Option<(String, Option<DateTime<Utc>>)> {
+    if let (Ok(nix_hash), maybe_nix_timestamp) = (
+        std::env::var("NIX_JJ_GIT_HASH"),
+        std::env::var("NIX_JJ_GIT_TIMESTAMP"),
```
this is missing `rerun-if-env-changed` https://doc.rust-lang.org/cargo/reference/build-scripts.html#rerun-if-env-changed

jyn514 replied:
you have it for `NIX_JJ_GIT_HASH` above, but not for `NIX_JJ_GIT_TIMESTAMP`. might make sense to move the println into this function, or add a wrapper which prints rerun every time you try to access the variable.

jyn514 commented on cli/build.rs:
```diff
@@ -41,21 +44,51 @@ fn main() -> std::io::Result<()> {
+    if let Some((mut git_hash, maybe_date)) = get_git_timestamp_and_hash() {
+        git_hash.truncate(16);
+        println!(
+            "cargo:rustc-env=JJ_VERSION={}+{}-{}",
+            version,
```
unrelated to this PR, but it would be simpler to use `CARGO_PKG_VERSION` instead of parsing the version from `cargo metadata` https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts (see "CARGO_PKG_<var>")
